File: TaskSheduler.py
import threading
import heapq
import datetime
import time
import queue


# ---------------- THREAD POOL ----------------
class ThreadPool:
    def __init__(self, num_workers):
        self.queue = queue.Queue()
        self.workers = []

        for _ in range(num_workers):
            t = threading.Thread(target=self._worker, daemon=True)
            t.start()
            self.workers.append(t)

    def submit(self, func, *args):
        self.queue.put((func, args))

    def _worker(self):
        while True:
            func, args = self.queue.get()
            try:
                func(*args)
            finally:
                self.queue.task_done()


# ---------------- TASK ----------------
class Task:
    def __init__(self, name, func, start_time, interval=None, executions=1):
        self.name = name
        self.func = func
        self.start_time = start_time
        self.interval = interval
        self.executions = executions

        # prevent overlap
        self.exec_lock = threading.Lock()

    def __lt__(self, other):
        return self.start_time < other.start_time


# ---------------- SCHEDULER ----------------
class TaskScheduler:
    def __init__(self, num_workers=4):
        self.lock = threading.Lock()
        self.condition = threading.Condition(self.lock)
        self.heap = []
        self.running = True

        self.thread_pool = ThreadPool(num_workers)

        self.scheduler_thread = threading.Thread(target=self._run, daemon=True)
        self.scheduler_thread.start()

    def schedule(self, task):
        with self.condition:
            heapq.heappush(self.heap, task)
            self.condition.notify()

    def _run(self):
        while self.running:
            with self.condition:
                while not self.heap:
                    self.condition.wait()

                task = self.heap[0]
                now = datetime.datetime.now()

                if task.start_time > now:
                    wait_time = (task.start_time - now).total_seconds()
                    self.condition.wait(timeout=wait_time)
                    continue

                heapq.heappop(self.heap)

            # -------- EXECUTION (via thread pool) --------
            if task.exec_lock.acquire(blocking=False):
                def run(t):
                    try:
                        t.func()
                    finally:
                        t.exec_lock.release()

                self.thread_pool.submit(run, task)

            # -------- RESCHEDULE --------
            if task.interval and task.executions > 1:
                task.executions -= 1
                task.start_time = datetime.datetime.now() + datetime.timedelta(seconds=task.interval)
                self.schedule(task)

    def stop(self):
        self.running = False
        with self.condition:
            self.condition.notify_all()


# ---------------- TEST ----------------
def work():
    print("Running at", datetime.datetime.now())
    time.sleep(2)


if __name__ == "__main__":
    scheduler = TaskScheduler(num_workers=3)

    # one-time task
    t1 = Task("task1", work, datetime.datetime.now() + datetime.timedelta(seconds=1))

    # recurring task
    t2 = Task(
        "task2",
        work,
        datetime.datetime.now() + datetime.timedelta(seconds=2),
        interval=3,
        executions=3
    )

    scheduler.schedule(t1)
    scheduler.schedule(t2)

    time.sleep(10)
    scheduler.stop()

File: TaskSheduler.java
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.locks.*;

// ---------------- TASK ----------------
class Task implements Comparable<Task> {
    String name;
    Runnable func;
    long runAt;          // epoch millis
    long interval;       // millis
    int executions;

    private final ReentrantLock execLock = new ReentrantLock();

    public Task(String name, Runnable func, long runAt, long interval, int executions) {
        this.name = name;
        this.func = func;
        this.runAt = runAt;
        this.interval = interval;
        this.executions = executions;
    }

    @Override
    public int compareTo(Task other) {
        return Long.compare(this.runAt, other.runAt);
    }

    public boolean tryLock() {
        return execLock.tryLock();
    }

    public void unlock() {
        execLock.unlock();
    }
}

// ---------------- SCHEDULER ----------------
class TaskScheduler {

    private final PriorityQueue<Task> heap = new PriorityQueue<>();
    private final ReentrantLock lock = new ReentrantLock();
    private final Condition condition = lock.newCondition();

    private final ExecutorService threadPool;
    private volatile boolean running = true;

    public TaskScheduler(int workers) {
        this.threadPool = Executors.newFixedThreadPool(workers);

        Thread schedulerThread = new Thread(this::run);
        schedulerThread.setDaemon(true);
        schedulerThread.start();
    }

    public void schedule(Task task) {
        lock.lock();
        try {
            heap.offer(task);
            condition.signal();
        } finally {
            lock.unlock();
        }
    }

    private void run() {
        while (running) {
            Task task;

            lock.lock();
            try {
                while (heap.isEmpty()) {
                    condition.await();
                }

                task = heap.peek();
                long now = System.currentTimeMillis();

                if (task.runAt > now) {
                    long waitTime = task.runAt - now;
                    condition.await(waitTime, TimeUnit.MILLISECONDS);
                    continue;
                }

                heap.poll();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                continue;
            } finally {
                lock.unlock();
            }

            // -------- EXECUTION --------
            if (task.tryLock()) {
                threadPool.submit(() -> {
                    try {
                        task.func.run();
                    } finally {
                        task.unlock();
                    }
                });
            }

            // -------- RESCHEDULE --------
            if (task.interval > 0 && task.executions > 1) {
                task.executions--;
                task.runAt = System.currentTimeMillis() + task.interval;
                schedule(task);
            }
        }
    }

    public void stop() {
        running = false;
        lock.lock();
        try {
            condition.signalAll();
        } finally {
            lock.unlock();
        }
        threadPool.shutdown();
    }
}

// ---------------- TEST ----------------
public class Main {
    public static void main(String[] args) throws InterruptedException {

        TaskScheduler scheduler = new TaskScheduler(3);

        Runnable work = () -> {
            System.out.println("Running at " + System.currentTimeMillis());
            try {
                Thread.sleep(2000);
            } catch (InterruptedException ignored) {}
        };

        // one-time task
        scheduler.schedule(new Task("task1", work,
                System.currentTimeMillis() + 1000, 0, 1));

        // recurring task
        scheduler.schedule(new Task("task2", work,
                System.currentTimeMillis() + 2000, 3000, 3));

        Thread.sleep(10000);
        scheduler.stop();
    }
}